src/mp_units.cpp

```diff
@@ -0,0 +1,74 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2018 Mateusz Pusz
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+module;
+
+#include <cassert>
+#include <gsl/gsl-lite.hpp>
+
+export module mp_units;
+
+export import <array>;
+export import <compare>;
+import <optional>;
+import <ratio>;
+import <tuple>;
+
+// core
+#include <units/base_dimension.h>
+#include <units/bits/unit_text.h>
+#include <units/customization_points.h>
+#include <units/derived_dimension.h>
+#include <units/exponent.h>
+#include <units/generic/dimensionless.h>
+#include <units/kind.h>
+#include <units/magnitude.h>
+#include <units/math.h>
+#include <units/point_origin.h>
+#include <units/prefix.h>
+#include <units/quantity.h>
+#include <units/quantity_cast.h>
+#include <units/quantity_kind.h>
+#include <units/quantity_point.h>
+#include <units/quantity_point_kind.h>
+#include <units/random.h>
+#include <units/ratio.h>
+#include <units/reference.h>
+#include <units/symbol_text.h>
+#include <units/unit.h>
+
+// systems
+#include <units/isq/dimensions.h>
+#include <units/isq/iec80000/iec80000.h>
+#include <units/isq/natural/natural.h>
+#include <units/isq/si/cgs/cgs.h>
+#include <units/isq/si/hep/hep.h>
+#include <units/isq/si/iau/iau.h>
+#include <units/isq/si/si.h>
+#include <units/isq/si/typographic/typographic.h>
+#include <units/isq/si/uscs/uscs.h>
```

Comment:
Is it possible to support submodules already?

Reply:
Yes. However, I decided against that. Up to `mp_units.core` and `mp_units.isq.dimensions` and perhaps some of the lighter systems, everything compiled under a second. But `mp_units.isq.si` and some others took more, bringing the total to almost 3 minutes.

The current single module `mp_units` takes almost 1 minute to compile. All times can be halved if I could find a way to compile the object and CMI in parallel while still requiring the CMI to recompile if the object needs to recompile, which is currently done as a direct dependency and thus are not parallelized.

Reply:
Why do separate modules compile 3 times slower than the one big module? Some things like IO will probably be defined as separate modules as it might be a no-go for some customers to use that.

Reply:
Some things like IO will probably have to be defined as separate modules as it might be a no-go for some customers to use that. Also, I would like to have at least core being separate from the systems as well (in case someone wants to just use the framework and not systems predefined by us).

Reply:
I don't know and can only speculate. Perhaps each has to do some amount of repeated instantiations that just doesn't happen in One Big Module because they're memoized.

Reply:
The double compilation is not required after the realization stated at https://github.com/mpusz/units/pull/350#discussion_r1126285956. I'll look into https://github.com/mpusz/units/pull/350#discussion_r841088509.

Reply:
I did some measurements.
Compiled item | Time
--- | ---
Module `mp_units` | 33 s
Static tests (`#include`) | 1 m 11 s
Static tests (`mp_units`) | 3 m 34 s

I compiled a target with the sources of the static tests in both modes (`#include`, `mp_units`)
with property `CXX_COMPILER_LAUNCHER` `"${CMAKE_COMMAND};-E;time"`.
For good measure, I used `-j1`.

Compiled item | `#include` time | Module `mp_units` time
--- | --- | ---
[  3%] fixed_string_test.cpp.o | Elapsed time: 1 s. (time), 0.000485 s. (clock) | Elapsed time: 4 s. (time), 0.00048 s. (clock)
[  6%] symbol_text_test.cpp.o | Elapsed time: 1 s. (time), 0.000436 s. (clock) | Elapsed time: 4 s. (time), 0.000775 s. (clock)
[ 10%] ratio_test.cpp.o | Elapsed time: 1 s. (time), 0.00061 s. (clock) | Elapsed time: 9 s. (time), 0.00046 s. (clock)
[ 13%] prime_test.cpp.o | Elapsed time: 1 s. (time), 0.000454 s. (clock) | Elapsed time: 1 s. (time), 0.000537 s. (clock)
[ 16%] type_list_test.cpp.o | Elapsed time: 2 s. (time), 0.000521 s. (clock) | Elapsed time: 11 s. (time), 0.00057 s. (clock)
[ 20%] dimension_op_test.cpp.o | Elapsed time: 2 s. (time), 0.000505 s. (clock) | Elapsed time: 12 s. (time), 0.000696 s. (clock)
[ 23%] chrono_test.cpp.o | Elapsed time: 4 s. (time), 0.000527 s. (clock) | Elapsed time: 16 s. (time), 0.00055 s. (clock)
[ 26%] custom_rep_test_min_expl.cpp.o | Elapsed time: 2 s. (time), 0.000508 s. (clock) | Elapsed time: 14 s. (time), 0.000586 s. (clock)
[ 30%] custom_rep_test_min_impl.cpp.o | Elapsed time: 3 s. (time), 0.000521 s. (clock) | Elapsed time: 16 s. (time), 0.000421 s. (clock)
[ 33%] custom_unit_test.cpp.o | Elapsed time: 6 s. (time), 0.000467 s. (clock) | Elapsed time: 15 s. (time), 0.000497 s. (clock)
[ 36%] dimensions_concepts_test.cpp.o | Elapsed time: 19 s. (time), 0.000459 s. (clock) | Elapsed time: 12 s. (time), 0.000589 s. (clock)
[ 40%] kind_test.cpp.o | Elapsed time: 3 s. (time), 0.00046 s. (clock) | Elapsed time: 1 s. (time), 0.000487 s. (clock)
[ 43%] references_test.cpp.o | Elapsed time: 3 s. (time), 0.000495 s. (clock) | Elapsed time: 14 s. (time), 0.000592 s. (clock)
[ 46%] si_test.cpp.o | Elapsed time: 22 s. (time), 0.000531 s. (clock) | Elapsed time: 20 s. (time), 0.000504 s. (clock)
[ 50%] unit_test.cpp.o | Elapsed time: 2 s. (time), 0.000612 s. (clock) | Elapsed time: 11 s. (time), 0.000525 s. (clock)
[ 53%] iec80000_test.cpp.o | Elapsed time: 3 s. (time), 0.000444 s. (clock) | Elapsed time: 14 s. (time), 0.000478 s. (clock)
[ 56%] cgs_test.cpp.o | Elapsed time: 5 s. (time), 0.000506 s. (clock) | Elapsed time: 16 s. (time), 0.000554 s. (clock)
[ 60%] si_cgs_test.cpp.o | Elapsed time: 7 s. (time), 0.000552 s. (clock) | Elapsed time: 18 s. (time), 0.000453 s. (clock)
[ 63%] fps_test.cpp.o | Elapsed time: 8 s. (time), 0.000485 s. (clock) | Elapsed time: 17 s. (time), 0.000415 s. (clock)
[ 66%] si_fps_test.cpp.o | Elapsed time: 13 s. (time), 0.000521 s. (clock) | Elapsed time: 18 s. (time), 0.000474 s. (clock)
[ 70%] si_hep_test.cpp.o | Elapsed time: 13 s. (time), 0.000592 s. (clock) | Elapsed time: 14 s. (time), 0.000544 s. (clock)
[ 73%] us_test.cpp.o | Elapsed time: 3 s. (time), 0.000477 s. (clock) | Elapsed time: 15 s. (time), 0.000514 s. (clock)
[ 76%] concepts_test.cpp.o | Elapsed time: 5 s. (time), 0.000466 s. (clock) | Elapsed time: 18 s. (time), 0.000449 s. (clock)
[ 80%] math_test.cpp.o | Elapsed time: 5 s. (time), 0.000456 s. (clock) | Elapsed time: 13 s. (time), 0.000471 s. (clock)
[ 83%] point_origin_test.cpp.o | Elapsed time: 5 s. (time), 0.000476 s. (clock) | Elapsed time: 1 s. (time), 0.000503 s. (clock)
[ 86%] angle_test.cpp.o | Elapsed time: 2 s. (time), 0.000873 s. (clock) | Elapsed time: 14 s. (time), 0.00044 s. (clock)
[ 90%] quantity_kind_test.cpp.o | Elapsed time: 9 s. (time), 0.000499 s. (clock) | Elapsed time: 23 s. (time), 0.000902 s. (clock)
[ 93%] quantity_point_kind_test.cpp.o | Elapsed time: 8 s. (time), 0.000565 s. (clock) | Elapsed time: 20 s. (time), 0.00084 s. (clock)
[ 96%] quantity_point_test.cpp.o | Elapsed time: 8 s. (time), 0.000472 s. (clock) | Elapsed time: 15 s. (time), 0.000584 s. (clock)
[100%] quantity_test.cpp.o | Elapsed time: 8 s. (time), 0.000499 s. (clock) | Elapsed time: 19 s. (time), 0.000785 s. (clock)

Reply:
WOW, that is not what I expected. I thought that the user's code would benefit from having all the instantiations done already in the module. But it seems that modules actually provide much worse performance to the users :-(

Reply:
`mp_units.core` is a win. But as mentioned at https://github.com/mpusz/units/pull/350#discussion_r840593083, there's little else I can do. Most systems require SI, which is heavy.

Compiled item | `#include` | `mp_units.core`
--- | --- | ---
`angle_test` | 2 s | 1 s
`dimension_op_test` | 2 s | 0 s
`fixed_string_test` | 1 s | 0 s
`prime_test` | 1 s | 1 s
`ratio_test` | 1 s | 0 s
`symbol_text_test` | 1 s | 0 s
`type_list_test` | 2 s | 0 s
`unit_test` | 2 s | 1 s

Reply:
But why preprocessing SI from scratch all over again in a header file is faster than reusing already parsed info in a module?

Reply:
My guess is that accessing what's been written to the PCM (Clang's CMI, Compiled Module Interface) is slower to access than what the compiler parses from textual inclusion, when the target PCM is a heavy template library, as in this case.

Reply:
Opened https://github.com/llvm/llvm-project/issues/61226.

Reply:
The LLVM issue seems fixed.